xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -553,6 +567,36 @@ public void run() {
       }
     }
 
+    private long generateHash(List<HashPolicy> hashPolicies, Map<String, String> headers) {
+      Long hash = null;
+      for (HashPolicy policy : hashPolicies) {
+        Long newHash = null;
+        if (policy.type() == HashPolicy.Type.HEADER) {
+          if (headers.containsKey(policy.headerName())) {
+            String value = headers.get(policy.headerName());
+            if (policy.regEx() != null && policy.regExSubstitution() != null) {
+              value = policy.regEx().matcher(value).replaceAll(policy.regExSubstitution());
+            }
+            newHash = hashFunc.hashAsciiString(value);
+          }
+        } else if (policy.type() == HashPolicy.Type.CHANNEL_ID) {
+          newHash = hashFunc.hashLong(logId.getId());
```

Comment:
A problem of using resolver id instead real channel id: what if channel goes to IDLE and out of IDLE, and the resolver gets restarted, a new resolver id is generated?

Reply:
As discussed elsewhere, affinity may not worth much for RPCs after Channel goes into/out of IDLE. See https://github.com/grpc/grpc/pull/25415#discussion_r587712169.